presto-main/src/main/java/com/facebook/presto/metadata/PolymorphicScalarFunctionBuilder.java

```diff
@@ -249,6 +223,56 @@ public MethodsGroup build()
         }
     }
 
+    public static class ChoiceBuilder
+    {
+        private final Class<?> clazz;
+        private Signature signature;
+        private boolean nullableResult;
+        private List<ArgumentProperty> argumentProperties = emptyList();
+        private List<MethodsGroup> methodsGroups = new ArrayList<>();
+
+        private ChoiceBuilder(Class<?> clazz, Signature signature)
+        {
+            this.clazz = clazz;
+            this.signature = requireNonNull(signature, "signature is null");
+        }
+
+        public ChoiceBuilder implementation(Function<MethodsGroupBuilder, MethodsGroupBuilder> methodGroupSpecification)
+        {
+            MethodsGroupBuilder methodsGroupBuilder = new MethodsGroupBuilder(clazz);
+            methodGroupSpecification.apply(methodsGroupBuilder);
+            MethodsGroup methodsGroup = methodsGroupBuilder.build();
+            methodsGroups.add(methodsGroup);
+            return this;
+        }
+
+        public ChoiceBuilder nullableResult(boolean nullableResult)
+        {
+            this.nullableResult = nullableResult;
+            return this;
+        }
+
+        public ChoiceBuilder argumentProperties(ArgumentProperty... argumentProperties)
+        {
+            this.argumentProperties = ImmutableList.copyOf(requireNonNull(argumentProperties, "argumentProperties is null"));
```

Comment:
I remember getting a comment in the past about not mixing non-null checks with other logic on the same line, so instead:

```
requireNonNull(argumentProperties, "argumentProperties is null");
this.argumentProperties = ImmutableList.copyOf(argumentProperties);
```